pkg/controllers/node/ipam.go

```diff
@@ -128,44 +339,22 @@ func (c *NodeController) gatherIPAMData() (map[string][]model.AllocationAttribut
 			attr := b.Attributes[*idx]
 
 			// Track nodes based on IP allocations.
-			if val, ok := attr.AttrSecondary[ipam.AttributeNode]; ok {
-				if _, ok := calicoNodes[val]; !ok {
-					calicoNodes[val] = []model.AllocationAttribute{}
-				}
-
+			if node, ok := attr.AttrSecondary[ipam.AttributeNode]; ok {
 				// Update metrics maps with this allocation.
-				allocationsByNode[val]++
-				if b.Affinity == nil || val != strings.TrimPrefix(*b.Affinity, "host:") {
+				if b.Affinity == nil || node != strings.TrimPrefix(*b.Affinity, "host:") {
 					// If the allocations node doesn't match the block's, then this is borrowed.
-					borrowedIPsByNode[val]++
+					borrowedIPsByNode[node]++
 				}
-
-				// If there is no handle, then skip this IP. We need the handle
-				// in order to release the IP below.
-				if attr.AttrPrimary == nil {
-					ip := ordinalToIP(b, *idx)
-					logc := log.WithFields(log.Fields{"ip": ip, "block": b.CIDR.String()})
-					if firstSkip {
-						logc.Warnf("Skipping IP with no handle")
-						firstSkip = false
-					} else {
-						logc.Debugf("Skipping IP with no handle")
-					}
-					continue
-				}
-
-				// Add this allocation to the node, so we can release it later if
-				// we need to.
-				calicoNodes[val] = append(calicoNodes[val], attr)
 			}
 		}
 	}
-	log.Debugf("Calico nodes found in IPAM: %v", calicoNodes)
 
 	// Update prometheus metrics.
 	ipsGauge.Reset()
-	for node, num := range allocationsByNode {
-		ipsGauge.WithLabelValues(node).Set(float64(num))
+	for node, allocations := range c.allocationsByNode {
+		if num := len(allocations); num != 0 {
```

Comment:
This was a corresponding change to match that the `allocationsByNode` map will now have entries for nodes with 0 allocations, whereas before it would have not had those entries. There was a test asserting behavior in that scenario so this change maintains how it used to behave (that nodes with 0 allocations don't show up in the metrics). 

I don't think I have a strong opinion on if that's the right behavior, just maintaining what the code used to do.

Reply:
With my latest changes this is no longer needed, as `allocationsByNode` no longer contains entries for nodes that have 0 allocations.